lib/logcat.js

```diff
@@ -1,118 +1,80 @@
-  this.proc = spawn(this.adb.path, this.adb.defaultArgs.concat(['logcat']));
-  this.proc.stdout.setEncoding('utf8');
-  this.proc.stderr.setEncoding('utf8');
-  this.proc.on('error', function (err) {
-    logger.error('Logcat capture failed: ' + err.message);
-    if (!this.calledBack) {
-      this.calledBack = true;
-      cb(err);
-    }
-  }.bind(this));
-  this.proc.on('exit', function (code, signal) {
-    logger.debug('Logcat terminated with code ' + code + ', signal ' + signal);
+class Logcat {
+  constructor (opts = {}) {
+    this.adb = opts.adb;
+    this.debug = opts.debug;
+    this.debugTrace = opts.debugTrace;
     this.proc = null;
-  }.bind(this));
-  this.proc.stdout.pipe(through(this.onStdout.bind(this)));
-  this.proc.stderr.pipe(through(this.onStderr.bind(this)));
-};
-
-Logcat.prototype.stopCapture = function (cb) {
-  logger.debug("Stopping logcat capture");
-  if (this.proc === null) {
-    logger.debug("Logcat already stopped");
-    cb();
-    return;
+    this.logs = [];
+    this.logsSinceLastRequest = [];
   }
-  this.proc.on('exit', function () {
-    cb();
-  });
-  this.proc.kill();
-  this.proc = null;
-};
 
-Logcat.prototype.onStdout = function (data) {
-  this.onOutput(data, '');
-};
-
-Logcat.prototype.onStderr = function (data) {
-  if (/execvp\(\)/.test(data)) {
-    logger.error('Logcat process failed to start');
-    if (!this.calledBack) {
-      this.calledBack = true;
-      this.onLogcatStart(new Error("Logcat process failed to start"));
-      return;
-    }
+  startCapture () {
+    return new Promise(async (resolve, reject) => {
+      log.debug("Starting logcat capture");
+      this.proc = new SubProcess(this.adb.path, this.adb.defaultArgs.concat(['logcat']));
+      await this.proc.start(0);
+      this.proc.on('exit', (code, signal) => {
+        log.info(`Logcat terminated with code ${code}, signal ${signal}`);
+        this.proc = null;
+      });
+      this.proc.on('lines-stderr', lines => {
+        for (let line of lines) {
+          if (/execvp\(\)/.test(line)) {
+            log.error('Logcat process failed to start');
+            reject(new Error(`Logcat process failed to start. stderr: ${line}`));
+          }
+          resolve();
+          this.formatOuput(line, ' STDERR');
+        }
+      });
+      this.proc.on('lines-stdout', lines => {
+        resolve();
+        for (let line of lines) {
+          this.formatOuput(line, '');
+        }
+      });
+    });
   }
-  this.onOutput(data, ' STDERR');
-};
 
-Logcat.prototype.onOutput = function (data, prefix) {
-  if (!this.logcatStarted) {
-    this.logcatStarted = true;
-    if (!this.calledBack) {
-      this.calledBack = true;
-      this.onLogcatStart();
-    }
-  }
-  data = data.trim();
-  data = data.replace(/\r\n/g, "\n");
-  var logs = data.split("\n");
-  _.each(logs, function (log) {
-    log = log.trim();
-    if (log) {
-      var logObj = {
+  formatOuput (output, prefix) {
```

Comment:
This method do more than format output it may be worth doing some cleanup there.